### PR TITLE
Adapt attribute completion from IPython completer

### PR DESCRIPTION
**Pull request by icholy**
I copied over the [attribute completion from IPython](https://github.com/ipython/ipython/blob/7e15d0696ab27d8b72dad6774d0fc378b4c546e5/IPython/core/completer.py#L354) and simplified it a bit.

### CODE REVIEW

**Comment by berkerpeksag** [comment]
Whoa, amazing! Could you fix this warning?

```
hy/completer.py:89:71: E202 whitespace before ']'
```

LGTM.

File: hy/completer.py [hunk 1]
```diff
@@ -64,7 +66,31 @@ def __init__(self, namespace=None):
 
         self.namespace = namespace
 
-    def complete(self, text, state):
+    def attr_matches(self, text):
+        "Borrowed from IPython's completer"
```
Comment (berkerpeksag) [inline hy/completer.py]: You could make this a `# comment`.

File: hy/completer.py [hunk 2]
```diff
@@ -27,6 +27,8 @@
 
 import os
 import sys
+import re
```
Comment (berkerpeksag) [inline hy/completer.py]: Please alphabetize imports.

File: hy/completer.py [hunk 3]
```diff
@@ -77,6 +103,20 @@ def complete(self, text, state):
             [matches.append(x) for x in
                 filter(lambda x: x.startswith(text), p)]
 
+        return matches
+
+    def complete(self, text, state):
+
```
Comment (berkerpeksag) [inline hy/completer.py]: Could you delete these empty lines?

File: hy/completer.py [hunk 4]
```diff
@@ -77,6 +103,20 @@ def complete(self, text, state):
...
+        if "." in text:
+            try:
+                matches = self.attr_matches(text)
+            except:
+                import traceback
+                traceback.print_exc()
+        else:
+            matches = self.global_matches(text)
         try:
             return matches[state]
```
Comment (berkerpeksag) [inline hy/completer.py]: So `matches` can be both a list and a dict, right? Could you add a comment about this?

File: hy/cmdline.py [hunk 5]
```diff
@@ -326,6 +330,7 @@ def cmdline_handler(scriptname, argv):
                 sys.exit(x.errno)
 
     # User did NOTHING!
+
```
Comment (berkerpeksag) [inline hy/cmdline.py]: Please remove this empty line too.

File: hy/cmdline.py [hunk 6]
```diff
@@ -219,9 +220,12 @@ def run_repl(hr=None, spy=False):
     sys.ps1 = "=> "
     sys.ps2 = "... "
 
-    with completion():
+    ns = dict()
```
Comment (berkerpeksag) [inline hy/cmdline.py]: Nitpick: `dict()` -> `{}`.

**Comment by berkerpeksag** [comment]
And please add your name to `AUTHORS` :)

**Comment by icholy** [comment]
I need to figure out the conversion between `_` and `-` still. I'll reopen when that's working.

File: hy/completer.py (hunk 4 above)
Comment (icholy) [inline hy/completer.py]: @berkerpeksag I don't think so?

Comment (berkerpeksag) [inline hy/completer.py]: Ignore my comment, I just saw `IndexError` :)

**Pull request closed without merging**